kinesis_consumer.cpp

```diff
@@ -0,0 +1,266 @@
+/*-------------------------------------------------------------------------
+ *
+ * kinesis_consumer.cpp
+ *	  Implementation for kinesis consumer C wrapper
+ *
+ * Copyright (c) 2013-2016, PipelineDB
+ *
+ *-------------------------------------------------------------------------
+ */
+#include <aws/kinesis/KinesisClient.h>
+#include <aws/core/client/ClientConfiguration.h>
+#include <aws/core/auth/AWSCredentialsProviderChain.h>
+#include <aws/kinesis/model/GetRecordsResult.h>
+#include <aws/kinesis/model/GetShardIteratorResult.h>
+#include <aws/kinesis/model/GetShardIteratorRequest.h>
+#include <aws/core/utils/Outcome.h>
+#include <aws/core/utils/logging/AWSLogging.h>
+#include <aws/core/utils/logging/DefaultLogSystem.h>
+#include <aws/kinesis/model/GetRecordsRequest.h>
+#include <aws/core/client/RetryStrategy.h>
+#include <aws/core/utils/logging/FormattedLogSystem.h>
+#include <aws/core/client/DefaultRetryStrategy.h>
+
+#include "kinesis_consumer.h"
+#include "conc_queue.hpp"
+
+#include <unordered_set>
+#include <functional>
+#include <thread>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/time.h>
+
+using namespace Aws::Auth;
+using namespace Aws::Client;
+using namespace Aws::Kinesis;
+using namespace Aws::Kinesis::Model;
+using namespace Aws::Utils;
+
+struct kinesis_consumer
+{
+	KinesisClient *client;
+	concurrent_queue<GetRecordsOutcome*> *queue;
+	std::atomic<bool> keep_running;
+	std::thread *thread;
+	Aws::String shard_iter;
```

Comment:
Each consumer is probably going to need some kind of list of shards that it is responsible for reading from.